### PR TITLE
syntax: Tweak path parsing logic

### PR DESCRIPTION
**Pull request by petrochenkov**
Associated paths starting with `<<` are parsed in patterns.

Paths like `self::foo::bar` are interpreted as paths and not as `self` arguments in methods (cc @matklad).
Now, I believe, _all_ paths are consistently parsed greedily in case of ambiguity.
Detection of `&'a mut self::` requires pretty large (but still fixed) lookahead, so I had to increase the size of parser's lookahead buffer.
Curiously, if `lookahead_distance >= lookahead_buffer_size` was used previously, the parser hung forever, I fixed this as well, now it ICEs.

r? @jseyfried

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -5961,10 +5968,8 @@ impl<'a> Parser<'a> {
                                     maybe_append(attrs, extra_attrs));
             return Ok(Some(item));
         }
-        if self.check_keyword(keywords::Union) &&
-                self.look_ahead(1, |t| t.is_ident() && !t.is_any_keyword()) {
+        if self.eat_keyword(keywords::Union) {
```
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: I think this could break a macro invocation `union!();` in an item position.

**Review by jseyfried — COMMENTED** [review]
Nice! r=me modulo `union!();`

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -245,6 +245,23 @@ enum PrevTokenKind {
...
 
+// Simple circular buffer used for keeping few next tokens.
+#[derive(Default)]
+struct LookaheadBuffer {
+    buffer: [TokenAndSpan; LOOKAHEAD_BUFFER_CAPACITY],
+    start: usize,
+    end: usize,
+}
+
+const LOOKAHEAD_BUFFER_CAPACITY: usize = 8;
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: `x % LOOKAHEAD_BUFFER_CAPACITY` is faster.
Not much difference either way, though.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -245,6 +245,23 @@ enum PrevTokenKind {
...
+
+impl LookaheadBuffer {
+    fn len(&self) -> usize {
+        let shift = if self.start <= self.end { 0 } else { LOOKAHEAD_BUFFER_CAPACITY };
+        shift + self.end - self.start
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Yep, looks simpler.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Hmm, right, will fix.

**Pull request merged**